common/functions/src/scalars/strings/string2number.rs

```diff
@@ -64,44 +61,53 @@ where
             features = features.monotonicity();
         }
 
-        FunctionDescription::creator(Box::new(Self::try_create)).features(features)
+        Function2Description::creator(Box::new(Self::try_create)).features(features)
     }
 }
 
-/// A common function template that transform string column into string column
-/// Eg: trim, lower, upper, etc.
-impl<T, R> Function for String2NumberFunction<T, R>
+/// A common function template that transform string column into number column
+/// Eg: length
+impl<T, R> Function2 for String2NumberFunction<T, R>
 where
-    T: NumberResultFunction<R> + Clone + Sync + Send,
-    R: DFPrimitiveType + Clone,
-    DFPrimitiveArray<R>: IntoSeries,
+    T: NumberOperator<R> + Clone,
+    R: PrimitiveType + Clone + ToDataType,
 {
     fn name(&self) -> &str {
-        self.display_name.as_str()
+        &self.display_name
     }
 
-    fn return_type(&self, args: &[DataTypeAndNullable]) -> Result<DataTypeAndNullable> {
-        if !args[0].is_numeric() && !args[0].is_string() && !args[0].is_null() {
-            return Err(ErrorCode::IllegalDataType(format!(
-                "Expected string or null, but got {}",
+    fn return_type(
+        &self,
+        args: &[&common_datavalues2::DataTypePtr],
+    ) -> Result<common_datavalues2::DataTypePtr> {
+        // We allow string, numeric AND null as input
+        if args[0].data_type_id().is_string()
+            || args[0].data_type_id().is_numeric()
+            || args[0].data_type_id() == TypeID::Null
+        {
+            Ok(R::to_data_type())
+        } else {
+            Err(ErrorCode::IllegalDataType(format!(
+                "Expected string, numeric or null, but got {:?}",
                 args[0]
-            )));
+            )))
         }
-        let nullable = args.iter().any(|arg| arg.is_nullable());
-        let dt = T::return_type()?;
-        Ok(DataTypeAndNullable::create(&dt, nullable))
     }
 
-    fn eval(&self, columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
-        let column = columns[0]
-            .column()
-            .cast_with_type(&DataType::String)?
```

Comment:
@sundy-li It seems that I need to input into `String`? I'm meeting errors like

```
2022-02-09T09:34:07.340102Z ERROR databend_query::servers::mysql::writers::query_result_writer: OnQuery Error: Code: 1058, displayText = downcast column error, column type: UInt64, expected column: "common_datavalues2::columns::string::StringColumn".
```

logs: https://github.com/datafuselabs/databend/runs/5122838220?check_suite_focus=true#step:3:1761

Reply:
Yes, I think we should forbid the implicit cast in `return_type`, so some tests should be removed.

1. return_type just check the args[0] is `string` or not.

2. `passthrough_null` is default to true, so we do not need to check null types.


In `function2_adaptor`:
```
  if self.passthrough_null() {
            // one is null, result is null
            if args.iter().any(|v| v.data_type_id() == TypeID::Null) {
                return Ok(NullType::arc());
            }
```